### DISCUSSION TITLE
Infinispan Error while trying to create a channel using the specified configuration with Keycloak 26.0.7

### DISCUSSION THREAD

**Discussion by ora7598** [keycloak/keycloak#36165] [Q&A]
Hi there!
I have recently deployed 2 pods running with Keycloak 26.0.7 on the same K8S Cluster + enabled distributed cache with a custom Infinispan config XML. 
It worked as expected until I needed to redeploy the KC (nothing has changed regarding the Infinispan/KC version, just some minor custom provider changes), and I got an error from Infinispan which for an unknown reason got fixed after a few restarts. But now it happened again:
`ISPN000541: Error while trying to create a channel using the specified configuration '[TCP(bundler.max_size=64000, sock_conn_timeout=300, linger=-1, thread_pool.keep_alive_time=60000, diag.enabled=false, bind_port=7800, thread_naming_pattern=pl, non_blocking_sends=false, thread_pool.thread_dumps_threshold=10000, send_buf_size=640k, thread_pool.max_threads=200, use_virtual_threads=false, bundler_type=transfer-queue, thread_pool.min_threads=0), RED(), JDBC_PING(insert_single_sql=INSERT INTO public.JGROUPSPING (own_addr, cluster_name, bind_addr, updated, ping_data) values (?, ?, '127.0.0.1', NOW(), ?), connection_driver=org.postgresql.Driver, delete_single_sql=DELETE FROM public.JGROUPSPING WHERE own_addr=? AND cluster_name=?, info_writer_sleep_time=500, remove_all_data_on_view_change=true, select_all_pingdata_sql=SELECT ping_data, own_addr, cluster_name FROM public.JGROUPSPING WHERE cluster_name=?, connection_password=<hidden-on-purpose>, connection_url=<hidden-on-purpose>, initialize_sql=CREATE SCHEMA IF NOT EXISTS public; CREATE TABLE IF NOT EXISTS public.JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, bind_addr varchar(200) NOT NULL, updated timestamp default current_timestamp, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name)), clear_sql=DELETE FROM public.JGROUPSPING WHERE cluster_name=?, connection_username=<hidden-on-purpose>), MERGE3(max_interval=30000, min_interval=10000), FD_SOCK2(offset=50000), FD_ALL3(), VERIFY_SUSPECT2(timeout=1000), pbcast.NAKACK2(xmit_interval=200, xmit_table_num_rows=50, resend_last_seqno=true, use_mcast_xmit=false, xmit_table_msgs_per_row=1024, xmit_table_max_compaction_time=30000), UNICAST3(conn_close_timeout=5000, xmit_interval=200, xmit_table_num_rows=50, xmit_table_msgs_per_row=1024, xmit_table_max_compaction_time=30000), pbcast.STABLE(desired_avg_gossip=5000, max_bytes=1M), pbcast.GMS(join_timeout=2000, print_local_addr=false), UFC(min_threshold=0.40, max_credits=4m), MFC(min_threshold=0.40, max_credits=4m), FRAG4(frag_size=60000)]'`

Custom Infinispan config XML file:
<?xml version="1.0" encoding="UTF-8"?>
<infinispan
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="urn:infinispan:config:15.0 http://www.infinispan.org/schemas/infinispan-config-15.0.xsd"
        xmlns="urn:infinispan:config:15.0">

    <jgroups>
        <stack name="jdbc-ping-tcp" extends="tcp">
            <JDBC_PING
                    connection_driver="org.postgresql.Driver"
                    connection_username="${env.KC_DB_USERNAME}"
                    connection_password="${env.KC_DB_PASSWORD}"
                    connection_url="${env.KC_DB_URL}"
                    initialize_sql="CREATE SCHEMA IF NOT EXISTS ${env.KC_DB_SCHEMA:public}; CREATE TABLE IF NOT EXISTS ${env.KC_DB_SCHEMA:public}.JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, bind_addr varchar(200) NOT NULL, updated timestamp default current_timestamp, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name))"
                    insert_single_sql="INSERT INTO ${env.KC_DB_SCHEMA:public}.JGROUPSPING (own_addr, cluster_name, bind_addr, updated, ping_data) values (?, ?, '${env.JGROUPS_DISCOVERY_EXTERNAL_IP:127.0.0.1}', NOW(), ?)"
                    delete_single_sql="DELETE FROM ${env.KC_DB_SCHEMA:public}.JGROUPSPING WHERE own_addr=? AND cluster_name=?"
                    select_all_pingdata_sql="SELECT ping_data, own_addr, cluster_name FROM ${env.KC_DB_SCHEMA:public}.JGROUPSPING WHERE cluster_name=?"
                    clear_sql="DELETE FROM ${env.KC_DB_SCHEMA:public}.JGROUPSPING WHERE cluster_name=?"
                    info_writer_sleep_time="500"
                    remove_all_data_on_view_change="true"
                    stack.combine="REPLACE"
                    stack.position="MPING"
            />
        </stack>
    </jgroups>

    <cache-container name="keycloak">
        <transport lock-timeout="60000" stack="jdbc-ping-tcp"/>
        <local-cache name="realms">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <local-cache name="users">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <distributed-cache name="sessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="authenticationSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="clientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineClientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="loginFailures" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <local-cache name="authorization">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="10000"/>
        </local-cache>
        <replicated-cache name="work">
            <expiration lifespan="-1"/>
        </replicated-cache>
        <local-cache name="keys">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <expiration max-idle="3600000"/>
            <memory max-count="1000"/>
        </local-cache>
        <distributed-cache name="actionTokens" owners="2">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <expiration max-idle="-1" lifespan="-1" interval="300000"/>
            <memory max-count="-1"/>
        </distributed-cache>
    </cache-container>
</infinispan>

**Comment by ahus1** [comment]
When running in Kubernetes with Keycloak 26.0, use the `kubernetes` stack for discovery. No XML configuration is needed, and we will be able to support you better.

With Keycloak 26.1, we still recommend to use `kubernetes` as the discovery mode on Kubernetes. Still, with KC 26.1, there is now a built-in configuration for JDBC Ping included which is the new default.

**Reply by ora7598** [reply]
Thanks @ahus1  !
I tried using cache-stack=kubernetes, but got the following errors:
`ISPN000541: Error while trying to create a channel using the specified configuration: 
'[TCP(bundler.max_size=64000, sock_conn_timeout=300, linger=-1, port_range │
│ =0, thread_pool.keep_alive_time=60000, diag.enabled=false, bind_port=7800, thread_naming_pattern=pl, non_blocking_sends=false, thread_pool.thread_dumps_threshold=10000, send_buf_size=640k, thre │
│ ad_pool.max_threads=200, use_virtual_threads=false, bundler_type=transfer-queue, logical_addr_cache_expiration=360000, thread_pool.min_threads=0), RED(), dns.DNS_PING(dns_record_type=A, num_dis │
│ covery_runs=3), MERGE3(max_interval=30000, min_interval=10000), FD_SOCK2(offset=50000), FD_ALL3(), VERIFY_SUSPECT2(timeout=1000), pbcast.NAKACK2(xmit_interval=200, xmit_table_num_rows=50, resen │
│ d_last_seqno=true, use_mcast_xmit=false, xmit_table_msgs_per_row=1024, xmit_table_max_compaction_time=30000), UNICAST3(conn_close_timeout=5000, xmit_interval=200, xmit_table_num_rows=50, xmit_t │
│ able_msgs_per_row=1024, xmit_table_max_compaction_time=30000), pbcast.STABLE(desired_avg_gossip=5000, max_bytes=1M), pbcast.GMS(join_timeout=2000, print_local_addr=false), UFC(min_threshold=0.4 │
│ 0, max_credits=4m), MFC(min_threshold=0.40, max_credits=4m), FRAG4(frag_size=60000)]`
 
And then another one:
`│ {"timestamp":"2025-01-16T14:23:39.865350863Z","sequence":9662,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler", │
│ "level":"ERROR","message":"ERROR: dns_query can not be null or empty","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"kc-1","}`

**Reply by ahus1** [reply]
[The docs state for the `kubernetes` stack the following](https://www.keycloak.org/server/caching#_available_transport_stacks): 

> DNS resolution using the JGroups DNS_PING protocol. It requires to set jgroups.dns.query to the headless service FQDN.

So you'll need to create a Kubernetes headless service for the Pods of Keycloak that you start up. That service name need to be passed as a property to the `JAVA_OPTS_APPEND` environment variable as described in the docs.

**Reply by ahus1** [reply]
BTW: If you have the chance to use the Keycloak Operator, this would configure this for you.

**Reply by ora7598** [reply]
Thanks @ahus1 !
I managed to run KC with `--cache-stack=kubernetes` (and I see all the "finished rebalance with members [x1, x2]" messages), but I'm facing an issue where the sessions do not seem to be synced or something..I have logged out from a session. Then when I tried to log in again I got a message that I couldn't log in, and I saw that the session was still active. Do you have any idea why this is happening?

**Reply by ahus1** [reply]
Happy to hear that the re-balancing now logs the right parts. 

To analyze why the login fails, I need more details beyond "I got a message that I couldn't log in". When there is a login error, there is usually a more elaborate error message in the log stating what failed and why. If you could share that, I might have more guidance.
